### PR TITLE
Dynatrace output plugin

### PR DESCRIPTION
**Pull request by thschue**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

File: plugins/outputs/dynatrace/dynatrace.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+	defer resp.Body.Close()
+
+	// print metric line results as info log
+	if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusAccepted {
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			d.Log.Errorf("Dynatrace error reading response")
+		}
+		bodyString := string(bodyBytes)
+		d.Log.Infof("Dynatrace returned: %s", bodyString)
```
Comment (ssoroka) [inline plugins/outputs/dynatrace/dynatrace.go:243]: Logger is too noisy. When working properly, almost nothing should be written to `Log.Info`. Probably suitable as a `Log.Debug`

File: plugins/outputs/dynatrace/dynatrace_test.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	d.EnvironmentURL = "test"
+	d.Log = testutil.Logger{}
+	err := d.Connect()
+	require.Error(t, err)
+}
+
+func TestSendMetric(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		json.NewEncoder(w).Encode(`{"linesOk":10,"linesInvalid":0,"error":null}`)
```
Comment (ssoroka) [inline plugins/outputs/dynatrace/dynatrace_test.go:96]: would be good to test the actual encoded data here to make sure it's formatting it properly. I'd like to see that for most of the outputs. Testing it connects isn't really enough.

File: plugins/outputs/dynatrace/dynatrace.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	for _, metric := range metrics {
+		// first write the tags into a buffer
+		tagb.Reset()
+		if len(metric.Tags()) > 0 {
+			for tk, tv := range metric.Tags() {
+				tagKey, err := d.normalize(tk, maxDimKeyLen)
+				if err != nil {
+					continue
+				}
+				fmt.Fprintf(&tagb, ",%s=%s", strings.ToLower(tagKey), d.escape(tv))
```
[the remark this answers is not in the document]
Comment (thschue) [reply plugins/outputs/dynatrace/dynatrace.go:156]: No, there are normalized values which are not necessarily lowercase.

File: plugins/outputs/dynatrace/dynatrace.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+	firstChars := true
+
+	for _, char := range s {
+		if firstChars && (unicode.IsDigit(char) || char == '_') {
+			continue
+		} else {
+			firstChars = false
+		}
+		normalizedString += string(char)
+	}
```
[the remark this answers is not in the document]
Comment (thschue) [reply plugins/outputs/dynatrace/dynatrace.go]: Changed

File: plugins/outputs/dynatrace/dynatrace.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+				}
+
+				metricID, err := d.normalize(d.Prefix+metric.Name()+"."+metricKey, maxMetricKeyLen)
+				// write metric name combined with its field
+				if err != nil {
+					continue
+				}
+				fmt.Fprintf(&buf, "%s", metricID)
+				// add the tag string
+				if len(tagb.String()) > 0 {
```
[the remark this answers is not in the document]
Comment (thschue) [reply plugins/outputs/dynatrace/dynatrace.go]: Changed

File: plugins/outputs/dynatrace/dynatrace.go (hunk 1 above)
Comment (thschue) [reply plugins/outputs/dynatrace/dynatrace.go]: Changed

File: plugins/outputs/dynatrace/dynatrace.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+		return fmt.Errorf("environmentApiToken is a required field for Dynatrace output")
+	}
+
+	d.client = &http.Client{
+		Transport: &http.Transport{
+			Proxy:           http.ProxyFromEnvironment,
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: d.SkipCertificateCheck},
+		},
+		Timeout: 5 * time.Second,
+	}
```
[the remark this answers is not in the document]
Comment (thschue) [reply plugins/outputs/dynatrace/dynatrace.go]: Changed

File: plugins/outputs/dynatrace/dynatrace.go [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+var (
+	reNameAllowedCharList = regexp.MustCompile("[^A-Za-z0-9.]+")
+	maxDimKeyLen          = 100
+	maxMetricKeyLen       = 250
+)
+
+// Dynatrace Configuration for the Dynatrace output plugin
+type Dynatrace struct {
+	EnvironmentURL       string          `toml:"environmentURL"`
+	EnvironmentAPIToken  string          `toml:"environmentApiToken"`
```
[the remark this answers is not in the document]
Comment (thschue) [reply plugins/outputs/dynatrace/dynatrace.go]: Changed

File: plugins/outputs/dynatrace/dynatrace_test.go (hunk 2 above)
Comment (wolfgangB33r) [reply plugins/outputs/dynatrace/dynatrace_test.go:129]: I added encoding tests within all outputs, according to your request.

**Comment by ssoroka** [comment]
I think we're in pretty good shape here now. There's maybe some extra checks that aren't necessary with the new TrimLeft call, but it's not a big deal. The sort in the tag names can be avoided by parsing the text into metrics in the test cases and using one of the methods like `testutil.MetricEqual` to compare them. 

Let me know when you're ready to merge and I'll get it in.

**Comment by wolfgangB33r** [comment]
@ssoroka I would rather prefer to sort them, as it keeps the output stable also within our ingest channel, which is preferable from my perspective. So from our side we are ready to merge. Thanks for all the help!

**Comment by ssoroka** [comment]
Thanks for the plugin! :D

**Pull request merged**